src/infill/ZigzagConnectorProcessor.cpp

```diff
@@ -0,0 +1,145 @@
+    //
+    // Decide whether we should add this connection or not.
+    // Add this zag connection is the following cases:
+    //  - if this zag lays in an even-numbered scanline segment
+    //  - if this zag is an end piece (check if the previous and the current scanlines are the same)
+    //    and "use end piece" is enabled
+    // Don't add a zag if:
+    //  - this zag is NOT an end piece, "skip some zags" is enabled, and this zag lays in a segment
+    //    which needs to be skipped.
+    // Moreover:
+    //  - if a "connected end pieces" is not enabled and this is an end piece, the last line
+    //    of this end piece will not be added.
+    //
+    // The rules above also apply to how the last part is processed (in polygon finishes)
+    //
+    const bool is_this_endpiece = start_scanline_idx == end_scanline_idx;
+    const bool is_this_connection_even = start_scanline_idx % 2 == 0;
+    bool should_skip_this_connection = false;
+    if (this->skip_some_zags && this->zag_skip_count > 0)
+    {
+        //
+        // Skipping some zags is done in the following way:
+        //  > direction: negative number means from <right> -> <left>
+        //               positive number means from <left> -> <right>
+        //  > for a connection:
+        //    - if it comes from <left> -> <right> (direction >= 0), we skip connections that
+        //      lay in segments which mean the condition: "index mod skip_count = 0"
+        //    - for connections that come from the other direction, we skip on
+        //      "(index - 1) mod skip_count = 0"
+        //
+        if (direction > 0)
+        {
+            should_skip_this_connection = start_scanline_idx % this->zag_skip_count == 0;
+        }
+        else
+        {
+            should_skip_this_connection = (start_scanline_idx - 1) % this->zag_skip_count == 0;
+        }
+    }
+
+    const bool should_add =
+        (is_this_connection_even && !is_this_endpiece && !should_skip_this_connection) // normal connections that should be added
+        || (this->use_endpieces && is_this_endpiece);  // end piece if it is enabled;
+
+    return should_add;
+}
+
+
+void ZigzagConnectorProcessor::registerScanlineSegmentIntersection(const Point& intersection, int scanline_index, int direction)
+{
+
+    if (this->is_first_connector)
+    {
+        // process as the first connector if we haven't found one yet
+        // this will be processed with the last remaining piece at the end (when the polygon finishes)
+        this->first_connector.push_back(intersection);
+        this->first_connector_index = scanline_index;
+        this->first_connector_direction = direction;
+        this->is_first_connector = false;
+    }
+    else
+    {
+        // add this connector if needed
+        if (this->shouldAddThisConnector(this->last_connector_index, scanline_index, direction))
+        {
+            for (unsigned int point_idx = 1; point_idx < this->current_connector.size(); ++point_idx)
```

Comment:
I think it's weird that this for loop starts from 1 and then in the loop we subtract 1 immediately, but this was in my original code, so I fixed this myself.